### ISSUE TITLE
Julia language server runs at 99% CPU indefinitely

### ISSUE THREAD

**Issue by quinnj** [JuliaEditorSupport/julia-vscode#201]
Unfortunately, I tried to setup Julia support in VS Code and upon updating the julia executable path to my Julia 0.6 binary, the language server process goes to 99% CPU indefinitely (until I kill it), confirmed by looking at Activity Monitor (on OSX). Any tips here? Anything I can load in the terminal to try and debug? This is my first attempt at using VS Code, so I really have no idea what's being attempted/loaded under the hood, so any direction would be helpful.

**Comment by ZacLN** [comment]
If you manually enable debug mode (change `ls_debug_mode` to `true` at https://github.com/JuliaEditorSupport/julia-vscode/blob/master/scripts/languageserver/main.jl#L26, where it's installed in your `.vscode/extensions/` directory). 

You can then see the messages send between the language server and editor by bringing up the output panel `View: Toggle Output` and switching to `julia Language Server` in the drop down on the rhs. If you paste the last message thats causing it to hang I'll try and work out whats

**Comment by davidanthoff** [comment]
Does that happen when you open any folder, or just for a particular folder, say a specific package?

**Comment by quinnj** [comment]
hmmmm, I just opened my JuliaLang/julia folder.

**Comment by ZacLN** [comment]
Ah, bingo. That'll a bit of a while to parse, especially if you have any previous builds stored (julia-__commitref__ dirs) there. We need to add a special case for editing `julia/base`

**Comment by quinnj** [comment]
oh, does it parse/evaluate every `.jl` file in the directory you open? it looks like it's trying to parse all `/base/*.jl` files. The other day, I also opened my `.julia` directory, so it was probably trying to parse every package source file I have. What's the idea here? Is that for auto-complete stuff? Is the recommended workflow to only open the direct folder you want to work on? And what's then recommended for working on Base? Also, is this the kind of overhead that happens every time I start up VS Code? Or is it a one-time parsing overhead and whatever info is cached somewhere?

**Comment by ZacLN** [comment]
Yep you've pretty much got it, it parses everything it can find in the open directory so it can connect files by `include` calls . 

The proposed workflow is to only open the project directory (or package directory) that you're working on. We've got no full setup for editing `Base` at the moment as you'd need the root directory to get `git` integration. I'll have a look tomorrow and see if I can set up a speedy special case for `Base`.

Our plan for the next major release is to introduce a background process to handle startup (and subsequent) parsing.

**Comment by quinnj** [comment]
Yeah, I think two things to explore would be some kind of incremental parsing, where it didn't have to all happen on startup. Or also caching, if you did a file hash like git on whatever you parse/index and then it could all be deserialized on startup. I'm fine paying a one-time parsing cost on Base if it means all subsequent startups are snappy and I get all the auto-complete goodness. Realistically, no one will be changing that many files in Base at once any way, so it shouldn't even be that heavy to invalidate a few files if they've been changed and reparse.

**Comment by davidanthoff** [comment]
JuliaLang/julia#22631 would be a god-send for this kind of problem: we could do all the parsing in a background thread while the LS still is responsive on another thread to any messages from the client. We have tinkered with various approaches to this by starting new julia processes that do some work in the background, but that is always cumbersome because one has to handle all the process management etc. manually.

In terms of caches, there are really two scenarios: one is that a user opens a project folder (say a package) and we need all the symbols from base and any packages, i.e. a cache of stuff from dependencies. The second scenario is the one @quinnj ran into: someone has a project folder that has a lot of .jl files in it, and we could cache that stuff. I think eventually both might be a good idea, but I think the first case is more pressing right now, while the second one is much more complicated (the cache needs to hold a lot more information than in the first case).

In general, caches are difficult, especially because we can't control how many LS processes will run at one point (and there can even be multiple LS processes that edit the same working folder). We did talk about a singleton symbol server process, that would solve the race condition problems in accessing the cache, but again, that adds a lot of complexity...

At the end of the day we probably want all of this, but some of this is quite tricky...

**Comment by ZacLN** [comment]
Re your first point on caches and dependencies: I think we came to the conclusion that a crashable worker process would be sufficient for actually loading the dependencies (maybe mixed in with a file watch to trigger a `reload` of altered packages).

For the problem of large projects (not incuding `Base`) we just need:
- A worker process handling parsing/linting in a non (main process) blocking fashion.
- An improved scheme for ordering file parsing (i.e. the currently open file, it's root file, then all the files included by the root, possibly by doing a quick regex for `r".jl"`

For `Base` I think we can just ignore parsing except for files that are open/edited and do some trickery of inserting the local edits into the Base/Core namespace.

**Comment by davidanthoff** [comment]
@ZacLN Should we put this onto the backlog for now, i.e. we won't fix this for the v0.7 release, right?

**Comment by ZacLN** [comment]
Yep, I'm going to change the title though
